v2: Use r.URL.RawPath for rewrite

When using e.g. `uri replace` paths get automatically unescaped. This might be undesirable. Something like this would fix it:

```diff
diff --git a/modules/caddyhttp/rewrite/rewrite.go b/modules/caddyhttp/rewrite/rewrite.go
index d47c388a..4370a77f 100644
--- a/modules/caddyhttp/rewrite/rewrite.go
+++ b/modules/caddyhttp/rewrite/rewrite.go
@@ -303,13 +303,24 @@ func (rep replacer) do(r *http.Request, repl *caddy.Replacer) bool {
        find := repl.ReplaceAll(rep.Find, "")
        replace := repl.ReplaceAll(rep.Replace, "")
 
-       oldPath := r.URL.Path
+       oldPath := r.URL.RawPath
        oldQuery := r.URL.RawQuery
 
-       r.URL.Path = strings.Replace(oldPath, find, replace, lim)
        r.URL.RawQuery = strings.Replace(oldQuery, find, replace, lim)
 
-       return r.URL.Path != oldPath && r.URL.RawQuery != oldQuery
+       if oldPath == "" {
+               oldPath = r.URL.Path
+               r.URL.Path = strings.Replace(oldPath, find, replace, lim)
+
+               return r.URL.Path != oldPath && r.URL.RawQuery != oldQuery
+       }
+
+       r.URL.RawPath = strings.Replace(oldPath, find, replace, lim)
+       if p, err := url.PathUnescape(r.URL.RawPath); err == nil {
+               r.URL.Path = p
+       }
+
+       return r.URL.RawPath != oldPath && r.URL.RawQuery != oldQuery
 }
 
 // Interface guard
```

What do you think? Maybe there's a better way?

Been fiddling with this for a while.

Unfortunately, the proposed fix makes the tests fail:

```
--- FAIL: TestRewrite (0.00s)
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:257: Test 38: Expected RequestURI='/foo/bar' but got '/'
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:260: Test 38: Expected URL='/foo/bar' but got ''
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:263: Test 38: Expected URL.RequestURI()='/foo/bar' but got '/'
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:257: Test 39: Expected RequestURI='/foo/replaced/bar' but got '/'
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:260: Test 39: Expected URL='/foo/replaced/bar' but got ''
    /home/matt/Dev/caddyserver/caddy/modules/caddyhttp/rewrite/rewrite_test.go:263: Test 39: Expected URL.RequestURI()='/foo/replaced/bar' but got '/'
```

--- a/modules/caddyhttp/rewrite/rewrite.go
+++ b/modules/caddyhttp/rewrite/rewrite.go
@@ -191,11 +191,21 @@ func (rewr Rewrite) rewrite(r *http.Request, repl *caddy.Replacer, logger *zap.L
 	// strip path prefix or suffix
 	if rewr.StripPathPrefix != "" {
 		prefix := repl.ReplaceAll(rewr.StripPathPrefix, "")
-		r.URL.Path = strings.TrimPrefix(r.URL.Path, prefix)
+		r.URL.RawPath = strings.TrimPrefix(r.URL.RawPath, prefix)
+		if p, err := url.PathUnescape(r.URL.RawPath); err == nil && p != "" {
+			r.URL.Path = p
+		} else {
+			r.URL.Path = strings.TrimPrefix(r.URL.Path, prefix)
+		}
 	}
 	if rewr.StripPathSuffix != "" {
 		suffix := repl.ReplaceAll(rewr.StripPathSuffix, "")
-		r.URL.Path = strings.TrimSuffix(r.URL.Path, suffix)
+		r.URL.RawPath = strings.TrimSuffix(r.URL.RawPath, suffix)
+		if p, err := url.PathUnescape(r.URL.RawPath); err == nil && p != "" {
+			r.URL.Path = p
+		} else {
+			r.URL.Path = strings.TrimSuffix(r.URL.Path, suffix)
+		}
 	}
 
 	// substring replacements in URI
@@ -289,10 +299,10 @@ type replacer struct {
 	Limit int `json:"limit,omitempty"`
 }
 
-// do performs the replacement on r and returns true if any changes were made.
-func (rep replacer) do(r *http.Request, repl *caddy.Replacer) bool {
+// do performs the replacement on r.
+func (rep replacer) do(r *http.Request, repl *caddy.Replacer) {
 	if rep.Find == "" || rep.Replace == "" {
-		return false
+		return
 	}
 
 	lim := rep.Limit
@@ -303,13 +313,14 @@ func (rep replacer) do(r *http.Request, repl *caddy.Replacer) bool {
 	find := repl.ReplaceAll(rep.Find, "")
 	replace := repl.ReplaceAll(rep.Replace, "")
 
-	oldPath := r.URL.Path
-	oldQuery := r.URL.RawQuery
+	r.URL.RawPath = strings.Replace(r.URL.RawPath, find, replace, lim)
+	if p, err := url.PathUnescape(r.URL.RawPath); err == nil && p != "" {
+		r.URL.Path = p
+	} else {
+		r.URL.Path = strings.Replace(r.URL.Path, find, replace, lim)
+	}
 
-	r.URL.Path = strings.Replace(oldPath, find, replace, lim)
-	r.URL.RawQuery = strings.Replace(oldQuery, find, replace, lim)
-
-	return r.URL.Path != oldPath && r.URL.RawQuery != oldQuery
+	r.URL.RawQuery = strings.Replace(r.URL.RawQuery, find, replace, lim)
 }
 
 // Interface guard

--- a/modules/caddyhttp/rewrite/rewrite_test.go
+++ b/modules/caddyhttp/rewrite/rewrite_test.go
@@ -201,6 +201,11 @@ func TestRewrite(t *testing.T) {
 		},
 		{
 			rule:   Rewrite{StripPathPrefix: "/prefix"},
+			input:  newRequest(t, "GET", "/prefix/foo%2Fbar"),
+			expect: newRequest(t, "GET", "/foo%2Fbar"),
+		},
+		{
+			rule:   Rewrite{StripPathPrefix: "/prefix"},
 			input:  newRequest(t, "GET", "/foo/prefix/bar"),
 			expect: newRequest(t, "GET", "/foo/prefix/bar"),
 		},
@@ -214,6 +219,11 @@ func TestRewrite(t *testing.T) {
 			rule:   Rewrite{StripPathSuffix: "suffix"},
 			input:  newRequest(t, "GET", "/foo/bar/suffix"),
 			expect: newRequest(t, "GET", "/foo/bar/"),
+		},
+		{
+			rule:   Rewrite{StripPathSuffix: "suffix"},
+			input:  newRequest(t, "GET", "/foo%2Fbar/suffix"),
+			expect: newRequest(t, "GET", "/foo%2Fbar/"),
 		},
 		{
 			rule:   Rewrite{StripPathSuffix: "/suffix"},
@@ -230,6 +240,11 @@ func TestRewrite(t *testing.T) {
 			rule:   Rewrite{URISubstring: []replacer{{Find: "findme", Replace: "replaced"}}},
 			input:  newRequest(t, "GET", "/foo/findme/bar"),
 			expect: newRequest(t, "GET", "/foo/replaced/bar"),
+		},
+		{
+			rule:   Rewrite{URISubstring: []replacer{{Find: "findme", Replace: "replaced"}}},
+			input:  newRequest(t, "GET", "/foo/findme%2Fbar"),
+			expect: newRequest(t, "GET", "/foo/replaced%2Fbar"),
 		},
 	} {
 		// copy the original input just enough so that we can

